### PR TITLE
fix(billing): restore payment type CRUD with CSRF-safe requests

### PR DESCRIPTION
**Pull request by yingbull**
Ontario payment-type create/edit/delete explicitly set the CSRF header that CSRFGuard also adds, producing duplicate values and a 403 instead of saving. The fix waits for token bootstrap, submits the token as a protected form parameter, retains automatic XHR protection, reports bootstrap failures visibly, and replaces unreadable `[object JSON]` AJAX error text. Delete no longer blocks the browser with synchronous XHR.

Fixes #3786. Adds a Playwright CRUD workflow with owned fixtures, literal-value checks, reopen checks and database assertions as part of #3774. Thirty-two production-JavaScript Node tests cover delayed/rejected token bootstrap, early clicks, bootstrap recovery for root and named contexts, duplicate request prevention, save/delete outcomes, error messages and retry.

The original defect is reproduced on the installed validation12 DEB in Ubuntu 26.04; header diagnostics counted duplicate segments without recording token values. The [validation15 installed-DEB report](https://github.com/carlos-emr/carlos/pull/3789#issuecomment-5742899511) records all four workflow steps passing. That run predates the early-click bootstrap fix, which has Node regression coverage. The latest follow-up reran local script checks; it did not repeat the installed-DEB run. No schema change.

## Summary by Sourcery

Restore reliable Ontario billing payment-type CRUD while preventing duplicate CSRF submission and expanding end-to-end and JavaScript regression coverage.

Bug Fixes:
- Restore Ontario billing payment-type create, edit, and delete operations by making their CSRF handling compatible with CSRFGuard.
- Improve payment-type save and delete failure reporting and allow users to retry after token or request failures.

Enhancements:
- Centralize payment-type CSRF bootstrap, token recovery, and request lifecycle handling while making deletion asynchronous.

Documentation:
- Document the payment-type CRUD coverage and installed-DEB validation results.

Tests:
- Add Playwright coverage for payment-type validation, create, edit, reopen, delete, and database persistence.
- Add production-JavaScript Node tests covering CSRF bootstrap timing and recovery, duplicate request prevention, response handling, and readable error reporting.

---
## Summary by cubic
Restores Ontario billing payment-type create/edit/delete, which returned 403 because the pages sent a `CSRF-TOKEN` header that duplicated the one CSRFGuard’s XHR wrapper adds. The pages now share `payment-type-csrf.js`: the token is submitted as a protected form parameter, bootstrap waits through `DOMContentLoaded`, pending requests are guarded, delete is asynchronous, and failures show readable alerts. A failed token bootstrap is retried by refetching the token from the context path in the script URL, including root deployments with an empty context prefix. Fixes #3786 as part of #3774. No schema change.

**Tests**
- Adds a Playwright CRUD workflow covering empty validation, create, edit, reopen, and delete with owned fixtures, database assertions, and literal quote/ampersand values.
- Adds Node tests executing the production JSP scripts for create/save/delete and the token helper, covering delayed or rejected bootstrap, early clicks before `DOMContentLoaded`, request gating, bootstrap recovery, delete failure reasons and retry, and save result handling.
- Validation of the latest code on an installed DEB is still pending.

<sup>Written for commit e9e6f0314346fd60d0d3a6b2bfaff967c13b7cae. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/carlos-emr/carlos/pull/3789?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by yingbull** [comment]
@coderabbitai full review

File: scripts/billing-payment-types.test.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    let resolve; let reject; let handler;
+    const ready = new Promise((yes,no) => { resolve = yes; reject = no; });
+    const requests = []; const alerts = [];
+    const tokenInput = { value: '' };
+    const doc = { querySelector: () => tokenInput };
+    function jquery() { return { ready: callback => callback(), on: (_event,_selector,callback) => { handler = callback; }, DataTable: () => {} }; }
+    jquery.ajax = request => requests.push(request);
+    const context = vm.createContext({ window: { csrfTokenReady: ready }, document: doc, jQuery: jquery, alert: text => alerts.push(text) });
+    const source = fs.readFileSync(path.join(root, 'manageBillingPaymentType.jsp'), 'utf8');
+    const code = [...source.matchAll(/<script\b[^>]*>([\s\S]*?)<\/script>/g)].map(match => match[1]).find(value => value.includes('data-paymentTypeId'));
```
[the remark this answers is not in the document]
Comment (yingbull) [reply scripts/billing-payment-types.test.js]: Fixed in bcc58359b6: the test-only inline script extractor now matches tag case insensitively. The actual production handlers run in fourteen passing Node cases, including delayed/failed CSRF initialization and explicit success/failure outcomes. The scanning thread is now resolved on the updated head.

File: scripts/billing-payment-types.test.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    let resolve; let reject; let handler;
+    const ready = new Promise((yes,no) => { resolve = yes; reject = no; });
+    const requests = []; const alerts = [];
+    const tokenInput = { value: '' };
+    const doc = { querySelector: () => tokenInput };
+    function jquery() { return { ready: callback => callback(), on: (_event,_selector,callback) => { handler = callback; }, DataTable: () => {} }; }
+    jquery.ajax = request => requests.push(request);
+    const context = vm.createContext({ window: { csrfTokenReady: ready }, document: doc, jQuery: jquery, alert: text => alerts.push(text) });
+    const source = fs.readFileSync(path.join(root, 'manageBillingPaymentType.jsp'), 'utf8');
+    const code = [...source.matchAll(/<script\b[^>]*>([\s\S]*?)<\/script>/gi)].map(match => match[1]).find(value => value.includes('data-paymentTypeId'));
```
[the remark this answers is not in the document]
Comment (yingbull) [reply scripts/billing-payment-types.test.js]: Addressed in a12b7dea1e. The test extractor now accepts whitespace and ignored attributes in closing script tags, with executable regression cases for both variants. All sixteen payment-script tests pass. This is extraction of trusted repository JSP source for tests, not an HTML sanitization boundary.

**Comment by yingbull** [comment]
@coderabbitai full review

File: scripts/billing-payment-types-playwright-checks.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  await link.first().click();
+  const page = await (await admin.locator('#dynamic-content iframe').elementHandle()).contentFrame();
+  h.assert(page, 'Payment management did not open its iframe');
+  const row = value => page.locator('#tblBillType tbody tr').filter({ hasText: value });
+  await s.step('empty payment type is rejected without a database write', async () => {
+    await page.getByRole('link', { name: 'Create a new payment type' }).click();
+    await page.locator('#paymentType').fill('');
+    const dialogs = await h.withExpectedDialogs(admin, () => page.locator('[name="create"]').click());
+    h.assert(dialogs.length === 1 && /can not be empty/.test(dialogs[0].text), 'Empty payment type did not show its validation alert');
+  });
```
[the remark this answers is not in the document]
Comment (yingbull) [reply scripts/billing-payment-types-playwright-checks.js:37]: Addressed in 1719573b08. The empty-input browser step now records the empty-row count before submission and verifies it is unchanged afterward. The updated live check is included in the validation15 DEB run.

File: src/main/webapp/billing/CA/ON/payment-type-csrf.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        alert("Success");
+        history.back();
+    } else {
+        alert(result && result.reason ? String(result.reason) : "Payment type was not saved.");
+    }
+}
+
+function paymentTypeRequestFailed(request, status, error) {
+    alert(status ? String(status) : error ? String(error) : "Unknown error happened!");
+}
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/main/webapp/billing/CA/ON/payment-type-csrf.js:144]: Addressed in 1719573b08. The shared error handler prefers responseJSON.reason, then the transport error, then HTTP status, with a readable fallback. Executable cases cover these precedence rules; all 18 payment JavaScript cases pass.

File: src/main/webapp/WEB-INF/jsp/billing/CA/ON/editBillingPaymentType.jsp [hunk 5]
```diff
@@ -59,78 +60,42 @@
...
-            return tokenInput ? tokenInput.value : "";
-        }
-
-        function createType() {
+        async function createType() {
             if (!check()) {
                 return;
             }
+            const token = await paymentTypeCsrfToken();
+            if (!token) return;
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/main/webapp/WEB-INF/jsp/billing/CA/ON/editBillingPaymentType.jsp:68]: Addressed in 1719573b08. Create, save and delete share an in-flight guard set before token retrieval and released on token failure or AJAX completion; requests have a bounded timeout. Tests exercise repeated activation during both bootstrap and pending AJAX, completion, and retry after bootstrap failure. All 18 cases pass; the updated live workflow is queued in the validation15 DEB run.

**Comment by yingbull** [comment]
Validation15 installed-DEB recheck passed all four payment workflow steps, including the new no-database-write assertion on empty input. All 18 executable JavaScript cases pass, including duplicate activation, retry and failure-message precedence. The three valid CodeRabbit findings are addressed in 1719573b08 and their threads resolved; a follow-up review is scheduled after the quota reset.

**Comment by yingbull** [comment]
@coderabbitai review

File: src/main/webapp/billing/CA/ON/payment-type-csrf.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
+/* Copyright (c) 2026 CARLOS Contributors. GPL-2.0-or-later. */
+// CSRFGuard supplies the XHR header; callers send this token in the form body only.
+async function paymentTypeCsrfToken() {
+    try {
+        if (window.csrfTokenReady) await window.csrfTokenReady;
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/main/webapp/billing/CA/ON/payment-type-csrf.js]: Valid, and fixed in 7d4eff05.

The buttons are parsed before `csrf-token.jspf`'s DOMContentLoaded listener has published the promise, so a click in that window skipped the wait, read the empty hidden input and showed the reload error for a bootstrap that was merely pending. `paymentTypeCsrfTokenReady()` now waits for `DOMContentLoaded` when `window.csrfTokenReady` is still null and `document.readyState === "loading"`, then awaits the promise. The continuation resumes as a microtask after the whole event dispatch, so the fragment's listener has installed the promise by then whichever order the two listeners were registered in.

Covered by a new Node case that drives the hostile listener order (helper registers first) and asserts the click waits, no alert fires, and the token resolves once bootstrap completes. All 19 payment-type cases and the full 790-case script suite pass.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by yingbull** [comment]
@coderabbitai full review

**Comment by yingbull** [comment]
@coderabbitai review

**Comment by yingbull** [comment]
Confirming DCO sign off for all commits at 61cdb41375868553227b41ccf70eab1d8c679a40

Review follow-up: verified the release merge in b2cf05879b preserves the payment-type, RH form, and BC billing workflow registrations. All six existing review threads are resolved, and their fixes remain present.

Addressed Copilot's remaining review-summary suggestion in 61cdb41375: eight new cases execute the production JSP delete callback for null/malformed responses, numeric/string application errors, explicit success, and transport failure. They verify rejection reasons, no redirect on failure, and retry after completion. All 27 payment-script cases pass, along with the suite-manifest and workflow-session checks; Playwright syntax and Ontario core registration also pass. No additional production defect found in this review.

Corrected the stale coverage note to reference the earlier validation15 installed-DEB success report. That report predates the early-click fix; this follow-up ran the local script checks, not a new installed-DEB rehearsal. CI is running on the new head.

**Comment by yingbull** [comment]
@coderabbitai full review

File: src/main/webapp/billing/CA/ON/payment-type-csrf.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+ * POST twice; a token failure releases the slot again so the user can retry.
+ *
+ * @returns {Promise<?string>} the token, or null when a request is already
+ *                             pending or the token is unavailable
+ */
+async function paymentTypeBeginRequest() {
+    if (paymentTypeRequestPending) return null;
+    paymentTypeRequestPending = true;
+    const token = await paymentTypeCsrfToken();
+    if (!token) paymentTypeRequestPending = false;
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/main/webapp/billing/CA/ON/payment-type-csrf.js:108]: Correct on both counts, and fixed in 4aac3065 by taking the refetch option.

A rejected `window.csrfTokenReady` stays rejected for the life of the document, so awaiting it again could only re-throw — releasing the guard bought nothing, and the first transient bootstrap failure bricked every later save until a reload. The existing retry case only covered an unpopulated input with no bootstrap promise, which is a different path.

`paymentTypeCsrfToken` now swallows the bootstrap rejection, and when no token is present calls `paymentTypeRefetchToken()`, which refetches once via `fetchCsrfToken`. This mirrors the retry already used by `efmformmanager.jsp:100`. Two details worth recording:

- A plain `.js` file cannot read `request.getContextPath()`, so the prefix `fetchCsrfToken` needs is trimmed off this script's own URL, captured at load time from `document.currentScript` (it is only set during first execution). Both pages serve it from `<contextPath>/billing/CA/ON/payment-type-csrf.js` — `editBillingPaymentType.jsp:49` and `manageBillingPaymentType.jsp:50` — so the suffix match is reliable, and it yields null rather than guessing if it ever stops holding.
- The call is guarded by `typeof fetchCsrfToken !== "function"` because `csrfTokenFetch.js` is loaded by `csrf-token.jspf` in the body, later than this file in the head. It is only reached on a click, by which time it is defined.

Three new cases cover a recovered bootstrap (asserting the derived context path and that no failure is reported), a refetch that also fails, and a page where no context path can be derived. 30 payment-script cases and the full 801-case script suite pass.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by yingbull** [comment]
@coderabbitai full review

**Comment by yingbull** [comment]
Addressed the additional Copilot review-summary comment (review 5257992703, root servlet context) in the latest `fix(billing): accept empty payment type context prefix` commit. The refetch guard now checks `paymentTypeContextPath === null`, preserving an empty string as a valid root prefix. Added executable cases for both the absolute root script URL browsers normally expose and a root-relative script URL yielding an empty prefix; all 32 payment-script tests pass.

For precision, the ordinary browser root deployment already derived an absolute origin from `document.currentScript.src`, so the previous falsy check did not reject that normal case. The explicit null check nevertheless makes the helper's unavailable-prefix contract correct and covers the empty-prefix case raised by Copilot. This finding was in a review summary rather than an inline thread; my earlier unresolved-thread check missed it.

**Comment by yingbull** [comment]
@coderabbitai full review
